dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/NeuralNetwork.java

```diff
@@ -205,29 +210,29 @@ public void train(final INDArray input, final int label) {
                                         final INDArray[] activations,
                                         final INDArray nextError) {
     if (begin == layers.length - 1) {
-      throw new RuntimeException("The beginning layer of backPropagateFromTo cannot be output layer");
+      throw new IllegalArgumentException("The beginning layer of backPropagateFromTo cannot be the output layer");
     }
     if (begin < end) {
-      throw new RuntimeException(String.format(
+      throw new IllegalArgumentException(String.format(
           "The beginning index (%d) must be greater than or equal to the ending index (%d).", begin, end));
     }
 
-    checkIndices(begin, end, true);
+    checkIndices(begin, end, false);
 
-    final INDArray[] errors = new INDArray[end - begin + 1];
+    final INDArray[] errors = new INDArray[begin - end + 1];
     INDArray error = nextError;
 
     for (int i = begin; i >= end; --i) {
-      error = layers[i].backPropagate(activations[i + 1], layers[i + 1].getLayerParameter(), error);
-      errors[i - begin] = error;
+      error = layers[i].backPropagate(activations[i], activations[i + 1], error);
```

Comment:
This changes which layer we're calling `getLayerParameter` on. Originally, we were using the *next* layer's parameters to compute the weight, but now we're using *this* layer's parameters. Could you explain in more detail what was changed?